Don't call cudaStreamDestroy at destruction time

dzhulgakov opened the pull request:
Summary:
It was leading to ocassional crashes with dynamically linked CUDA because runtime was already destroyed.

Also, unique_ptr<T[]> is more suitable than deque<T> for the purpose.

Differential Revision: D13571988

Yangqing commented on c10/cuda/CUDAStream.cpp:
```diff
@@ -18,9 +18,19 @@ namespace {
 // Internal implementation is entirely hidden
```
nit: this is a useless comment stating the obvious... see comment on line 33.

Yangqing commented on c10/cuda/CUDAStream.cpp:
```diff
@@ -18,9 +18,19 @@ namespace {
 // Internal implementation is entirely hidden
 struct CUDAStreamInternals {
   CUDAStreamInternals() = default;
+  CUDAStreamInternals(const CUDAStreamInternals&) = delete;
```
nit: should we use a macro for this to avoid potentially missing ctors/assignments?

Yangqing commented on c10/cuda/CUDAStream.cpp:
```diff
@@ -18,9 +18,19 @@ namespace {
+    // to a crash. It's likely an issue in CUDA, but to be safe - let's just
+    // "forget" the destruction.
+
+    // if (stream) cudaStreamDestroy(stream);
```
I am actually concerned that the name of the struct does not state that it's leaking streams. once the struct is destroyed the stream handle is forever lost till the end of the program lifetime. Should we, in line 18, clearly state that this is a leaky struct that is only supposed to be used inside one single place (aka holder for global streams inside this file)? Or better, name it LeakyCudaStreamInternals or something explicit.
I've seen cases when people carelessly move abstractions across files, and in the end get such tribal knowledge lost.

Yangqing commented on c10/cuda/CUDAStream.cpp:
```diff
@@ -43,7 +53,7 @@ static constexpr unsigned int kDefaultFlags = cudaStreamNonBlocking;
 // Default streams
 static std::once_flag init_flag;
-static std::vector<CUDAStreamInternals> default_streams;
+static std::unique_ptr<CUDAStreamInternals[]> default_streams;
```
If this is a static global variable, unique_ptr is probably not necessary especially if we know the compile time max default streams one will use (like, 16 as max number of gpus?)
Also, note that std::unique_ptr is not trivially destructible and thus is discouraged to be used in global variables (c.f. https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables)

Yangqing commented on c10/cuda/CUDAStream.cpp:
```diff
@@ -53,11 +63,16 @@ static std::vector<CUDAStreamInternals> default_streams;
+//
+// unique_ptr<T[]> is used instead of vector<T> because T might be non-moveable
+// and non-copyable.
+static std::unique_ptr<std::once_flag[]> device_flags;
```
ditto - raw arrays might do as well.

Yangqing commented on c10/cuda/CUDAStream.cpp:
```diff
@@ -175,12 +195,12 @@ static void initGlobalStreamState() {
   num_gpus = device_count();
```
Actually after seeing this I am pretty confident that we should use a compile time max GPU count to avoid the std::unique_ptr<x[]> complexity and use static arrays directly. Realistically one is not going to be using more than 16 GPUs and it saves a lot of potential pitfalls in the general C++ programming sense.

Yangqing reviewed (changes requested):
Thanks for the fix!

zdevito commented on c10/cuda/CUDAStream.cpp:
```diff
@@ -53,11 +61,16 @@ static std::vector<CUDAStreamInternals> default_streams;
+// unique_ptr<T[]> is used instead of vector<T> because T might be non-moveable
+// and non-copyable.
+static std::once_flag device_flags[C10_COMPILE_TIME_MAX_GPUS];
+static std::atomic<uint32_t> low_priority_counters[C10_COMPILE_TIME_MAX_GPUS];
```
You probably do not want locks used from different CPU threads on the same cache line.

Commit:
Don't call cudaStreamDestroy at destruction time (#15692)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/15692

It was leading to ocassional crashes with dynamically linked CUDA because runtime was already destroyed.

Also, unique_ptr<T[]> is more suitable than deque<T> for the purpose.

Reviewed By: Yangqing

Differential Revision: D13571988

fbshipit-source-id: 49e9cef1a4e59b8bc7d52474a71c95e44fe2bbbf
